Objects/bytesobject.c

```diff
@@ -3558,3 +3560,148 @@ _PyBytes_Repeat(char* dest, Py_ssize_t len_dest,
     }
 }
 
+/** Dedent a UTF-8 encoded string.
+ * behavior is expected to match `textwrap.dedent`
+ *
+ * return value:
+ * 0, no need to dedent, writer untouched
+ * 1, success
+ * -1, failure
+ *
+ * str is the beginning of the string to dedent.
+ * expecting (str != NULL)
+ *
+ * len is the length of the string to dedent.
+ * expecting (len >= 0)
+ *
+ * writer is a _PyBytesWriter object to write the dedented string.
+ * expecting (writer != NULL)
+ *
+ * p points to a char* indicating the current position in the _PyBytesWriter.
+ * It is updated to the new position after writing the dedented string on exit.
+ * expecting (p != NULL && *p != NULL)
+ */
+int
+_PyBytes_Dedent(const char *str, Py_ssize_t len, _PyBytesWriter *writer,
+                char **p)
```

Comment:
Now I think `_PyBytesWriter` is overkill here.
`char **out, Py_ssize_t *out_len` would be simple and better.